[storage] Introduce `store`

clabby opened the pull request:
## Overview

Introduces a new `store` module that contains a simple, unauthenticated `Store` K/V DB.

closes https://github.com/commonwarexyz/monorepo/issues/1206

patrick-ogrady commented:
👀 👀 👀

clabby commented on storage/src/store/operation.rs:
No changes here, but open to notes on the move. While this PR doesn't go all the way there, it seems like it would be nice to retroactively base the authenticated DB primitives off of the more simple unauthenticated `Store`. Glad to move this + `UpdateResult` back to `adb` if that's not the direction you guys are going to head with this, though 😄

patrick-ogrady replied:
This is exactly the decision point that came up during my review 😅 : https://github.com/commonwarexyz/monorepo/pull/1327#pullrequestreview-3059955917

I was originally thinking more or less "LSM DB replacement (without sorting)" for this PR (requiring vairable-size values) but am now pondering if your approach makes more sense (and if I really had any use cases in mind that had variable-sized data).

patrick-ogrady replied:
If we went the "LSM DB" route, I'd suggest reverting any changes in `adb` (but would be in support of this change if we opted to migrate `adb` to it).

patrick-ogrady commented on storage/src/store/mod.rs:
```diff
@@ -0,0 +1,79 @@
+//! A simple append-only key-value store backed by a log of state-change operations.
+//!
+//! This store provides append-only functionality without requiring MMR (Merkle Mountain Range) for
```
nit: I prefer describing `store` as-is (without any relation to MMRs) and then adding a recommendation to use `adb` if you want to authenticate data (the same could be done in reverse in the `adb/mod.rs` for "just need a store (without overhead of maintaining proofs)?"

https://github.com/commonwarexyz/monorepo/blob/31c6abdf1a264beb727817cdd20b34295b3a3fe4/storage/src/journal/fixed.rs#L7

patrick-ogrady commented on storage/src/store/mod.rs:
```diff
@@ -0,0 +1,79 @@
+//! authentication. It's designed for efficient local storage scenarios where cryptographic proofs
+//! are not required, making it more performant than [`crate::adb::Any`] for non-authenticated use cases.
+//!
+//! # Terminology
```
I think it is worth discussing `commit` in this section (as it supports rolling back to the last commit boundary).

patrick-ogrady commented on storage/src/store/base.rs:
```diff
@@ -0,0 +1,845 @@
+
+/// A simple append-only key-value store backed by a log of operations.
+///
+/// The [`Store`] provides persistent key-value storage with transactional semantics and automatic
```
nit: Again, prefer describing `store` as a standalone thing rather than a lighter adb::any (and adb::any as a more powerful version of this, which is more primitive).

patrick-ogrady commented on storage/src/store/base.rs:
```diff
@@ -0,0 +1,845 @@
+    /// The [`Store`]'s log of operations.
+    ///
+    /// This log is expected to be pruned to the `pruned_to_loc` boundary and contain all subsequent operations.
+    pub(super) log: Journal<E, Fixed<K, V>>,
```
I had originally intended for the `store` to not be integrated into `adb` (and use a `Variable<K, V>` here) but migrating `adb` now sounds tempting...

patrick-ogrady commented on storage/src/store/base.rs:
```diff
@@ -0,0 +1,845 @@
+        // TODO: Make the frequency with which we prune known inactive items configurable in case
+        // this turns out to be a significant part of commit overhead, or the user wants to ensure
+        // the log is backed up externally before discarding.
+        self.prune_inactive().await
```
In the office, we discussed making the actual `prune` call explicit (rather than running automatically).

Because we aren't syncing anything from `store`, this is ok but we'd need to offer some additional configuration if we want to integrate into `adb::any` (as we need to keep some before the inactive boundary to support dynamic state sync).

patrick-ogrady reviewed (commented):
My original ambition with this issue (#1206) was to actually create a totally unrelated implementation to `adb` (with variable-sized operations)...something like `archive` but "mutable" (and where each key wasn't associated with some "index").

This PR has made me second guess that original effort given how straightforward your abstraction seems to integrate into the `adb`s (especially because it is a "known useful" pattern rather than just something I thought may be useful).

The code you've provided looks pretty solid (I just left comments primarily around how it is presented).

Thanks for putting this up!

roberto-bayardo commented:
If decoupled from ADB, we don't need to support by-index lookups, and could therefore support variable length items without the additional log structure used by adb::immutable.

I'm leaning towards keeping this "non authenticated" store independent of ADB and giving it variable-length record capability to make it more flexible?

patrick-ogrady commented:
> If decoupled from ADB, we don't need to support by-index lookups, and could therefore support variable length items without the additional log structure used by adb::immutable.
> 
> I'm leaning towards keeping this "non authenticated" store independent of ADB and giving it variable-length record capability to make it more flexible?

Agreed. Let's keep store as a standalone thing for now (at least until we get further along the adb optimization path). We should support variable length as well (you can look at `archive::prunable` for an example of how this is done with `journal::variable` and `Index`).

clabby commented:
Sounds good - thanks y'all 😄 

I should have some time mid-week to revisit.

clabby commented:
Rewrote + simplified the store - now allows for variable-length values and is more separate from the `adb` in functionality.

patrick-ogrady commented on storage/src/store/mod.rs:
```diff
@@ -0,0 +1,619 @@
+        };
+
+        // Iterate over conflicts in the snapshot.
+        while cursor.next().is_some() {
```
If the `cursor` is `Some`, you'll need to iterate over all conflicts to determine if a conflict is a matching key (and update it accordingly): https://github.com/commonwarexyz/monorepo/blob/65ee8a7532a698b2a45bcfdc509f42c4fd49dfac/storage/src/adb/any/mod.rs#L447-L463

This approach may lead to you adding multiple instances of the same key to the `snapshot` (and `get` iteration will likely surface the first instance rather than the current). Outside of that issue, the size of the index will grow as a function of total ops (rather than active keys).

patrick-ogrady replied:
If the `cursor` is `Some`, you'll need to iterate over all conflicts to determine if a conflict is a matching key (and update it accordingly): https://github.com/commonwarexyz/monorepo/blob/65ee8a7532a698b2a45bcfdc509f42c4fd49dfac/storage/src/adb/any/mod.rs#L447-L463

This approach may lead to you adding multiple instances of the same key to the `snapshot` (and `get` iteration will likely surface the first instance rather than the current). Outside of that issue, the size of the index will grow as a function of total ops (rather than active keys).

IIRC this was correctly implemented in an earlier version of this PR but may be wrong. You should be able to write a test that causes unexpected behavior by calling `get` on such a key.

patrick-ogrady commented on storage/src/store/mod.rs:
```diff
@@ -0,0 +1,619 @@
+    ///
+    /// If the key does not exist, or the key has never taken on `version + 1` values, returns `Ok(None)`.
+    pub async fn get_historical(&self, key: &K, version: usize) -> Result<Option<V>, Error> {
+        if let Some(location) = self.snapshot.get(key).nth(version) {
```
The `snapshot` groups keys by truncated prefix, so the `nth` item may have nothing to do with the key you are querying by. Thus, the "nth" version of some key is really the `nth` version of some super set of keys with the same prefix...I think it would be better to remove this API because I don't think it makes sense to expose the user to this sort of inner complexity (done for efficiency).

patrick-ogrady commented on storage/src/store/mod.rs:
```diff
@@ -0,0 +1,619 @@
+    ///
+    /// If the key does not exist, returns `Ok(None)`.
+    pub async fn get_current(&self, key: &K) -> Result<Option<V>, Error> {
+        self.get_historical(key, self.get_value_count(key).saturating_sub(1))
```
As discussed in the comment on `get_historical`, this is not guaranteed to return the current value of a key (if it exists).

patrick-ogrady reviewed (changes requested):
Your "historical functionality" introduced some unintended behavior (based on a smal misunderstanding of how `Index` works).

If you revert those, I think it should be pretty close!

_Adding some sort of historical query pattern is an interesting idea (and as you are likely well-aware is commonly used to power archive state stores these days) but have typically found that to require some sort of "range query" functionality that isn't present in any of our DBs (so far 😉 )._

clabby commented:
Thanks @patrick-ogrady - good catch! I fooled myself on the underlying behavior of `Index` here with `test_store_historical_access`; The test only works out as-is due to using the same key, but "the "nth" version of some key is really the nth version of some super set of keys with the same prefix" (https://github.com/commonwarexyz/monorepo/pull/1327#discussion_r2249458713) bears its teeth when you assign a few values to a key that shares a prefix with the one used to saturate `values` as you suggested.

The code for the historical retrieval + allowing the index to expand with the number of ops was added in https://github.com/commonwarexyz/monorepo/pull/1327/commits/5be44916ff2971ff94f11413349ef91b2f0f81ad, and prior to that change, the store does handle conflicts within the cursor correctly (as suggested in https://github.com/commonwarexyz/monorepo/pull/1327#discussion_r2249458109.)

The downside to that version is that it does not support a form of pruning inactive operations in the log, which I attempted to give purpose to in https://github.com/commonwarexyz/monorepo/pull/1327/commits/5be44916ff2971ff94f11413349ef91b2f0f81ad by adding the option to access historical values that the key has taken on.

Would you prefer to keep it simple, and just revert to https://github.com/commonwarexyz/monorepo/pull/1327/commits/a3a3e559d7e3b8b1ff8a4fc6c9ae705caab86a5f, or is there more functionality you'd like to see here?

patrick-ogrady commented:
> Thanks @patrick-ogrady - good catch! I fooled myself on the underlying behavior of Index here with test_store_historical_access; The test only works out as-is due to using the same key, but "the "nth" version of some key is really the nth version of some super set of keys with the same prefix" (https://github.com/commonwarexyz/monorepo/pull/1327#discussion_r2249458713) bears its teeth when you assign a few values to a key that shares a prefix with the one used to saturate values as you suggested.

👍 

> The code for the historical retrieval + allowing the index to expand with the number of ops was added in https://github.com/commonwarexyz/monorepo/commit/5be44916ff2971ff94f11413349ef91b2f0f81ad, and prior to that change (https://github.com/commonwarexyz/monorepo/commit/a3a3e559d7e3b8b1ff8a4fc6c9ae705caab86a5f), the store does handle conflicts within the cursor correctly (as suggested in https://github.com/commonwarexyz/monorepo/pull/1327#discussion_r2249458109). The downside to that version is that it does not support a form of pruning inactive operations in the log, which I attempted to give purpose to in https://github.com/commonwarexyz/monorepo/commit/5be44916ff2971ff94f11413349ef91b2f0f81ad by adding the option to access historical values that the key has taken on. Would you prefer to keep it simple, and just revert to https://github.com/commonwarexyz/monorepo/commit/a3a3e559d7e3b8b1ff8a4fc6c9ae705caab86a5f, or is there more functionality you'd like to see here (i.e. pruning)?

Like with `adb::any`, we should maintain an "inactivity floor" and rewrite keys to the tip of the log (reading from the base of the log at the rate new keys are added to bound the "junk" on disk): https://github.com/commonwarexyz/monorepo/blob/65ee8a7532a698b2a45bcfdc509f42c4fd49dfac/storage/src/adb/any/mod.rs#L744-L793

You may definitely feel you are duplicating a good bit of logic and (as discussed previously) I'm ok with that for this early phase. We are planning to do a "deduplication effort" once the core storage primitives are complete (and prefer keeping them un-entangled for faster experimentation).

Hope that helps!

clabby commented:
Reverted the changes in https://github.com/commonwarexyz/monorepo/commit/5be44916ff2971ff94f11413349ef91b2f0f81ad, and added pruning logic to the `store` akin to `adb::any`'s.

patrick-ogrady commented on storage/src/store/mod.rs:
```diff
@@ -0,0 +1,822 @@
+    /// The operation is immediately visible in the snapshot for subsequent queries, but remains
+    /// uncommitted until [`Store::commit`] is called. Uncommitted operations will be rolled back
+    /// if the store is closed without committing.
+    pub async fn update(&mut self, key: K, value: V) -> Result<UpdateResult, Error> {
```
On mobile and haven't taken a full look (so ignore if wrong) but I think you're missing a delete op?

clabby replied:
Yeah - the `Store` currently doesn't support deletions, only overriding existing keys. Would you like that feature as well here?

https://github.com/commonwarexyz/monorepo/blob/aa42c41252d2348e509ff16b826058222a6a9035/storage/src/store/mod.rs#L5-L7

clabby commented on storage/src/store/mod.rs:
```diff
@@ -0,0 +1,990 @@
+    ///
+    /// Returns the number of operations that were applied to the store, the oldest retained
+    /// location, and the inactivity floor location.
+    async fn build_snapshot_from_log(mut self) -> Result<Self, Error> {
```
This method of building the snapshot is very memory inefficient at the moment; need to update this to trim committed operations as we stream the log's contents.

patrick-ogrady replied:
(As you originally called out, I definitely think there is a world where many of the adb's use store behind the scenes but we can make those migrations in subsequent PRs)

clabby reviewed (commented):
Added deletion support + fixed an issue with conflicting keys in the snapshot building process, one last todo.

clabby commented on storage/src/store/mod.rs:
```diff
@@ -0,0 +1,986 @@
+    /// uncommitted until [`Store::commit`] is called. Uncommitted operations will be rolled back
+    /// if the store is closed without committing.
+    pub async fn update(&mut self, key: K, value: V) -> Result<UpdateResult, Error> {
+        // Check if this is a no-op
+        if let Some(current_value) = self.get(&key).await? {
+            if current_value == value {
+                return Ok(UpdateResult::NoOp);
+            }
+        }
```
Curious y'all's opinion on whether or not this is desired. Removing this idea of no-op'ing `update` calls when the key has already assumed the same value results in a slightly more expensive pruning process due to unconditionally appending of a new operation, but removes an extra read from the locations map + op log in the `update` path.

patrick-ogrady replied:
Should get around to reviewing this sometime tomorrow 👍

patrick-ogrady commented on storage/src/store/mod.rs:
```diff
@@ -0,0 +1,919 @@
+    /// The [`Translator`] used by the compressed index.
+    pub translator: T,
+
+    /// The buffer pool to use for caching data.
```
I wonder if it worth added a TODO in `//` below this comment (or in the init function) to use this for the variable journal too?

Hopefully, we can get this in soon: https://github.com/commonwarexyz/monorepo/issues/1223

patrick-ogrady commented on storage/src/store/mod.rs:
```diff
@@ -0,0 +1,919 @@
+
+        // Append the operation to the entry log, and the offset to the locations log.
+        let (offset, _size) = self.log.append(current_section, op).await?;
+        self.locations.append(offset.into()).await?;
```
This is a clever optimization to only store offset here. I think it deserves a little additional documentation as to why we do this (using the "op number" to derive the section later).

clabby replied:
Added

patrick-ogrady commented on storage/src/store/mod.rs:
```diff
@@ -0,0 +1,919 @@
+        self.uncommitted_ops += 1;
+
+        // Append the operation to the entry log, and the offset to the locations log.
+        let (offset, _size) = self.log.append(current_section, op).await?;
```
nit: we use `_` for unused vars elswhere

patrick-ogrady commented on storage/src/store/mod.rs:
```diff
@@ -0,0 +1,919 @@
+    async fn apply_op(&mut self, op: Operation<K, V>) -> Result<u32, Error> {
+        let current_section = self.op_count / self.log_items_per_section;
+
+        self.uncommitted_ops += 1;
```
I think we should add this after the writes complete successfully.

clabby replied:
Done

patrick-ogrady commented on storage/src/store/mod.rs:
```diff
@@ -0,0 +1,919 @@
+
+/// Errors that can occur when interacting with a [`Store`] database.
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
```
nit: move this to the top of the file.

patrick-ogrady commented on storage/src/store/operation.rs:
```diff
@@ -0,0 +1,220 @@
+
+impl<K: Array, V: Codec> EncodeSize for Operation<K, V> {
+    fn encode_size(&self) -> usize {
+        match self {
```
nit: `1 + match self {`

(remove the +1 from each)

patrick-ogrady commented on storage/src/store/operation.rs:
```diff
@@ -0,0 +1,220 @@
+
+/// An operation applied to an unauthenticated database with a variable size value.
+#[derive(Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug)]
+pub enum Operation<K: Array, V: Codec> {
```
Please provide similar descriptions for each op: https://github.com/commonwarexyz/monorepo/blob/18e6449f12f57d4e2b3ae852be6676d24e204e08/storage/src/adb/operation.rs#L38-L46

(I like set/delete/commit -> i.e. what you have for this)

patrick-ogrady commented on storage/src/store/mod.rs:
```diff
@@ -0,0 +1,919 @@
+//! An unauthenticated key-value database supporting variable-sized values.
```
A mutable key-value database that supports variable-sized values.

patrick-ogrady commented on storage/src/store/mod.rs:
```diff
@@ -0,0 +1,919 @@
+//! 2. **Insertion**: Use [`Store::update`] to set a value for a given key
+//! 3. **Queries**: Use [`Store::get`] to retrieve current values
+//! 4. **Deletions**: Use [`Store::delete`] to remove a key's value
+//! 5. **Persistence**: Call [`Store::commit`] to make changes durable
```
Suggest moving `queries` below persistence.

patrick-ogrady commented on storage/src/store/mod.rs:
```diff
@@ -0,0 +1,919 @@
+//! tailing commit op. During this process, any encountered active operations are re-applied to the tip of
+//! the log.
+//!
+//! |                               Log State                                | Inactivity Floor | Uncommitted Ops |
```
Recommend adding a test that exactly replicates this case to ensure your counts are right.

patrick-ogrady reviewed (changes requested):
~5 minutes away from merge!
